kinesis/kinesis.go

```diff
@@ -0,0 +1,311 @@
+        if aerr, ok := err.(awserr.Error); ok {
+            if aerr.Code() == kinesis.ErrCodeProvisionedThroughputExceededException {
+                logrus.Warnf("[kinesis %d] Throughput limits for the delivery stream may have been exceeded.", outputPlugin.PluginID)
+                // Backoff and Retry
+                outputPlugin.backoff.StartBackoff()
+            }
+        }
+        return err
+    }
+    logrus.Debugf("[kinesis %d] Sent %d events to Kinesis\n", outputPlugin.PluginID, len(outputPlugin.records))
+
+    return outputPlugin.processAPIResponse(response)
+}
+
+// processAPIResponse processes the successful and failed records
+// it returns an error iff no records succeeded (i.e.) no progress has been made
+func (outputPlugin *OutputPlugin) processAPIResponse(response *kinesis.PutRecordsOutput) error {
+    if aws.Int64Value(response.FailedRecordCount) > 0 {
+        // start timer if all records failed (no progress has been made)
+        if aws.Int64Value(response.FailedRecordCount) == int64(len(outputPlugin.records)) {
+            outputPlugin.timer.Start()
+            return fmt.Errorf("PutRecords request returned with no records successfully recieved")
+        }
+
+        logrus.Errorf("[kinesis %d] %d records failed to be delivered\n", outputPlugin.PluginID, aws.Int64Value(response.FailedRecordCount))
+        failedRecords := make([]*kinesis.PutRecordsRequestEntry, 0, aws.Int64Value(response.FailedRecordCount))
+        // try to resend failed records
+        for i, record := range response.Records {
+            if record.ErrorMessage != nil {
+                logrus.Debugf("[kinesis %d] Record failed to send with error: %s\n", outputPlugin.PluginID, aws.StringValue(record.ErrorMessage))
+                failedRecords = append(failedRecords, outputPlugin.records[i])
+            }
+            if aws.StringValue(record.ErrorCode) == kinesis.ErrCodeProvisionedThroughputExceededException {
+                // Backoff and Retry
+                outputPlugin.backoff.StartBackoff()
+            }
+        }
+
+        outputPlugin.records = outputPlugin.records[:0]
+        outputPlugin.records = append(outputPlugin.records, failedRecords...)
+        outputPlugin.dataLength = 0
+        for _, record := range outputPlugin.records {
+            outputPlugin.dataLength += len(record.Data)
+        }
+    } else {
+        //request fully succeeded
+        for i, record := range response.Records {
+            logrus.Debugf("[kinesis %d] record- %d:Shard ID: %s", outputPlugin.PluginID, i, aws.StringValue(record.ShardId))
+        }
+        outputPlugin.timer.Reset()
+        outputPlugin.backoff.Reset()
+        outputPlugin.records = outputPlugin.records[:0]
+        outputPlugin.dataLength = 0
+    }
+    return nil
+}
+
+//randomString generates a random string of length 8
+//it uses the math/rand library
+func (outputPlugin *OutputPlugin)randomString() string {
+    b := make([]byte, 16)
```

Comment:
Why 16?

With 8 characters and your current charset, there are ~200 trillion possible partition keys.

Reply:
Also... this bit of code here could be made a lot more efficient.

This function could be called for every record- so you should allocate the array once and then store it some where. 

Allocating memory every time (and then later garbage collecting it) makes this slow.

Reply:
You could create a new struct which holds the seededRandom and the slice, and make it a field within the OutputPlugin struct. Separating the data makes the code cleaner.

Reply:
The length is set to be 8. Yes, we can make it better.

Reply:
Well, after removing this local allocation, I declared the buffer as the field of our output plugin. Sirprisingly, it performs a little bit slower instead of being faster. Still, I prefer to keep our new design as it seems more clean to me.